File: Alkitab/src/main/java/yuku/alkitab/base/util/ShareUrl.java
package yuku.alkitab.base.util;

import android.app.Activity;
import androidx.annotation.Keep;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.afollestad.materialdialogs.MaterialDialog;
import com.google.gson.JsonParseException;
import java.io.IOException;
import java.util.concurrent.atomic.AtomicBoolean;
import okhttp3.Call;
import okhttp3.FormBody;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import yuku.alkitab.base.App;
import yuku.alkitab.base.connection.Connections;
import yuku.alkitab.debug.BuildConfig;
import yuku.alkitab.debug.R;
import yuku.alkitab.model.Version;
import yuku.alkitab.util.Ari;
import yuku.alkitab.util.IntArrayList;

public class ShareUrl {
    public interface Callback {
        void onSuccess(String shareUrl);
        void onUserCancel();
        void onError(Exception e);
        void onFinally();
    }

    public static void make(@NonNull final Activity activity, final boolean immediatelyCancel, @NonNull final String verseText, final int ari_bc, @NonNull final IntArrayList selectedVerses_1, @NonNull final String reference, @NonNull final Version version, @Nullable final String preset_name, @NonNull final Callback callback) {
        if (immediatelyCancel) { // user explicitly ask for not submitting url
            callback.onUserCancel();
            callback.onFinally();
            return;
        }

        final StringBuilder aris = new StringBuilder();

        for (int i = 0, len = selectedVerses_1.size(); i < len; i++) {
            final int verse_1 = selectedVerses_1.get(i);
            final int ari = Ari.encodeWithBc(ari_bc, verse_1);
            if (aris.length() != 0) {
                aris.append(',');
            }
            aris.append(ari);
        }
        final FormBody.Builder form = new FormBody.Builder()
            .add("verseText", verseText)
            .add("aris", aris.toString())
            .add("verseReferences", reference);

        if (preset_name != null) {
            form.add("preset_name", preset_name);
        }

        final String versionLongName = version.getLongName();
        if (versionLongName != null) {
            form.add("versionLongName", versionLongName);
        }

        final String versionShortName = version.getShortName();
        if (versionShortName != null) {
            form.add("versionShortName", versionShortName);
        }

        final Call call = Connections.getOkHttp().newCall(
            new Request.Builder()
                .url(BuildConfig.SERVER_HOST + "v/create")
                .post(form.build())
                .build()
        );

        // when set to true, do not call any callback
        final AtomicBoolean done = new AtomicBoolean();

        final MaterialDialog dialog = new MaterialDialog.Builder(activity)
            .content("Getting share URL…")
            .progress(true, 0)
            .negativeText(R.string.cancel)
            .onNegative((dialog1, which) -> {
                if (done.getAndSet(true)) return;
                done.set(true);

                callback.onUserCancel();
                try {
                    dialog1.dismiss();
                } catch (Exception ignored) {
                }

                callback.onFinally();
            })
            .dismissListener(dialog1 -> {
                if (done.getAndSet(true)) return;

                callback.onUserCancel();
                try {
                    dialog1.dismiss();
                } catch (Exception ignored) {
                }

                callback.onFinally();
            })
            .show();

        call.enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(@NonNull final Call call, @NonNull final IOException e) {
                if (!done.getAndSet(true)) {
                    onComplete(() -> callback.onError(e));
                }
            }

            @Override
            public void onResponse(@NonNull final Call call, @NonNull final Response response) {
                if (done.getAndSet(true)) return;

                final ResponseBody body = response.body();
                if (body == null) {
                    onComplete(() -> callback.onError(new IOException("empty response body")));
                } else {
                    try {
                        final ShareUrlResponseJson obj = App.getDefaultGson().fromJson(body.charStream(), ShareUrlResponseJson.class);
                        if (obj.success) {
                            onComplete(() -> callback.onSuccess(obj.share_url));
                        } else {
                            onComplete(() -> callback.onError(new Exception(obj.message)));
                        }
                    } catch (JsonParseException e) {
                        onComplete(() -> callback.onError(e));
                    }
                }
            }

            void onComplete(@NonNull final Runnable todo) {
                activity.runOnUiThread(() -> {
                    if (activity.isFinishing()) return;

                    todo.run();

                    try {
                        dialog.dismiss();
                    } catch (Exception ignored) {
                    }

                    callback.onFinally();
                });
            }
        });
    }

    @Keep
    static class ShareUrlResponseJson {
        public boolean success;
        public String message;
        public String share_url;
    }
}

File: Alkitab/src/main/java/yuku/alkitab/base/util/ShareUrl.kt
package yuku.alkitab.base.util

import android.app.Activity
import androidx.annotation.Keep
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.callbacks.onDismiss
import com.google.gson.JsonParseException
import java.io.IOException
import java.util.concurrent.atomic.AtomicBoolean
import okhttp3.Call
import okhttp3.FormBody
import okhttp3.Request
import okhttp3.Response
import yuku.alkitab.base.App
import yuku.alkitab.base.connection.Connections
import yuku.alkitab.base.widget.MaterialDialogProgressHelper.progress
import yuku.alkitab.debug.BuildConfig
import yuku.alkitab.debug.R
import yuku.alkitab.model.Version
import yuku.alkitab.util.Ari
import yuku.alkitab.util.IntArrayList

object ShareUrl {
    fun make(
        activity: Activity,
        immediatelyCancel: Boolean,
        verseText: String,
        ari_bc: Int,
        selectedVerses_1: IntArrayList,
        reference: String,
        version: Version,
        preset_name: String?,
        callback: Callback,
    ) {
        if (immediatelyCancel) { // user explicitly ask for not submitting url
            callback.onUserCancel()
            callback.onFinally()
            return
        }

        val aris = (0 until selectedVerses_1.size())
            .map { Ari.encodeWithBc(ari_bc, selectedVerses_1[it]) }
            .joinToString(",")

        val form = FormBody.Builder().apply {
            add("verseText", verseText)
            add("aris", aris)
            add("verseReferences", reference)
            preset_name?.let { add("preset_name", it) }
            version.getLongName()?.let { add("versionLongName", it) }
            version.getShortName()?.let { add("versionShortName", it) }
        }.build()

        val call = Connections.okHttp.newCall(
            Request.Builder()
                .url(BuildConfig.SERVER_HOST + "/v/create")
                .post(form)
                .build()
        )

        // when set to true, do not call any callback
        val done = AtomicBoolean()

        val dialog = MaterialDialog(activity).show {
            message(text = "Getting share URL…")
            progress(true, 0)
            negativeButton(R.string.cancel) { dialog ->
                if (done.getAndSet(true)) return@negativeButton
                done.set(true)
                callback.onUserCancel()
                dialog.dismiss()
                callback.onFinally()
            }
            onDismiss { dialog ->
                if (done.getAndSet(true)) return@onDismiss
                callback.onUserCancel()
                dialog.dismiss()
                callback.onFinally()
            }
        }

        call.enqueue(object : okhttp3.Callback {
            override fun onFailure(call: Call, e: IOException) {
                if (!done.getAndSet(true)) {
                    onComplete { callback.onError(e) }
                }
            }

            override fun onResponse(call: Call, response: Response) {
                if (done.getAndSet(true)) return
                val body = response.body
                if (body == null) {
                    onComplete { callback.onError(IOException("empty response body")) }
                } else {
                    try {
                        val obj = App.getDefaultGson().fromJson(body.charStream(), ShareUrlResponseJson::class.java)
                        if (obj.success) {
                            onComplete { callback.onSuccess(obj.share_url) }
                        } else {
                            onComplete { callback.onError(Exception(obj.message)) }
                        }
                    } catch (e: JsonParseException) {
                        onComplete { callback.onError(e) }
                    }
                }
            }

            fun onComplete(todo: Runnable) {
                activity.runOnUiThread {
                    if (activity.isFinishing) return@runOnUiThread
                    todo.run()
                    try {
                        dialog.dismiss()
                    } catch (ignored: Exception) {
                    }
                    callback.onFinally()
                }
            }
        })
    }

    interface Callback {
        fun onSuccess(shareUrl: String)
        fun onUserCancel()
        fun onError(e: Exception)
        fun onFinally()
    }

    @Keep
    data class ShareUrlResponseJson(
        val success: Boolean,
        val message: String,
        val share_url: String,
    )
}
